### DISCUSSION TITLE
About sessions and windows and workflows

### DISCUSSION THREAD

**Discussion by unphased** [joshmedeski/sesh#67] [General]
Hi Josh, I watched your video on T which helped me immensely to understand your workflow. I'm a longtime tmux user and alacritty has also been my macOS terminal of choice. I am also an fzf enthusiast. I came upon this repo from a link from reddit, but it simply would not click for me, because the way I use tmux has been to very very seldom use sessions. So the notion of this deep streamlining around session navigation is very foreign to me but still very intriguing as it's likely to open up more flexibility if added to my workflows.

As part of transitioning from a vimL init script to a lua one as well as switching from vim to nvim, I found with the new plugins I am using (Heirline, Trouble, Nvim-tree) it has a very streamlined flow for switching buffers. As such I found I no longer regularly use vim tabs and mainly switch buffers now. What I found was that although tabs and splits are powerful, in practice work involves focusing on one file at a time, and what would repeatedly happen is the layouts I'd be building would keep getting messed up from the rapid navigation I'd do and it turns out more streamlined to mostly do away with having that layout state to begin with.  

It is clear that switching from using tmux windows and split panes to making use of sessions represents a similar transition at the tmux level, because I am having a similar inefficiency by continuing with my existing split-heavy tmux window usage. 

Since I often attempt to use terminals from really constrained interfaces like my phone, having a mostly full screen workflow like yours looks to benefit me greatly in that scenario when it comes to navigation. It also stands to streamline and improve focus in my usual environment with a bigger screen, so I'm excited to try this out. 

I still want to pick your brain though. 

- You don't use tmux with any "chrome" aka status lines. You just have a minimal title line at the top with the session name it seems. I can't tell if this is a statusline put at the top somehow, or a pane status line, in which case when you split panes in your session, all panes would then have a statusline indicating the same session name (that may be "inefficient").
- Why do you prefer it this way? Just minimalism? My 5+ year old approach to cramming all the information I figured might be useful (and some of it often is, like PID of child of the process! actual PID is usually just for the shell) is pretty great, but when my terminal is sufficiently small (like on a phone especially) it often crowds out the information I want, like the window bar, etc. Clearly it's time for me to make a change, and I need to take a page out of your book here and go minimalist too. If I were to do this though I would still use a status bar showing the current session name, but I would at minimum add a counter to show how many sessions total exist as a hint to go clean up once there are too many. But I would clean up the bulk of the crap I have in my status bar such as pane_title, pane_current_command, pane_current_path, etc.

- Def looks like you use sessions like I use my tmux windows. It actually seems like there is not much practical difference between these. I can think of at least one thing, with windows you can move panes around the windows. Not sure if it's possible to move a pane across sessions. I mean.. I never saw you open more tmux windows or do a single tmux split. Do you ever use these features? I think I will move more toward what you do and organize "workspaces" (regardless of whether they're primarily sessions or windows) by working dir, as that's what I found myself converging on organization by... but sometimes a workflow requires operations between two locations and side by side panes is the best way for that. Surely you have those too, in such a situation (say you have a session with a split in it open to terms in different dirs), when you are done with that, do you ... Nuke the session? Clean it up manually immediately? Leave it and clean it up later when you want to work in that dir again? Something else? Based on what I saw nuking it right away seems likely the way to go, since you have implemented super streamlined opening of a clean session on a dir name not-already-having-an-open-session.

- Although I enjoy fzf greatly and it is maximally efficient to go to a destination by inputting a portion of a name, When I work on stuff even 1s taken to type 5 chars to toggle back and forth is still much slower than simply switching via some next/prev navigation. For example the way I do things now is I'll usually have a tmux window for vim config and one for my current project I'm hacking on, and another for a different project. So, 3 windows. To switch between hacking on these 3 different things all I gotta do is cycle forward or back in the list of windows and I can see what they are and their order in the statusline. Similarly I'll always have at least 4 panes open in each window (in retrospect this is too much and i want to reduce it) and I have binds (directional ctrl+hjkl as well as other keys to cycle forward/back) to spatially switch across the panes.
-  From your video, the sense that I get is you have probably 5 or 6 sessions you'll regularly switch between and you're pretty committed to typing out the name each time to go there, with help from both zoxide and fzf. Now I'm not totally understanding how these two are combined together yet since they still feel like mutually exclusive tools to go from terse input to picking a real absolute directory path, but I'm sure I'll understand that soon enough. I did see you open up the tmux built in session switcher interface, which IMO is a fine and dandy interface. I think that this presents a linear ordering of your open sessions and we could hypothetically make a little statusline or some indicator for that state and have binds to cycle forward and back through sessions themselves. I feel like I would NEED this if I switch to maining session instead of windows. What do you think? Wouldn't you at least want a quick "go to prev open session" keybind? Some stuff at least to get to parity with for example OS native Cmd+Tab and Cmd+` for quickly switching to previous apps/windows?

**Comment by unphased** [comment]
oh ok, i get what fzf and zoxide are independently for. so zoxide is used like a shell working dir history manager here, also doubles as replacement for cd with some pixie dust sprinkled on top, got to use it at least a little to populate that history. fzf used as frontend, what's going on is feeding zoxide' list into fzf to give it an effectively auto-curated list of dirs to hang tmux sessions off of. 

to try to summarize my questioning a bit... i think all i'm thinking this system needs to fully round it out is a minimum surprise interface to rapidly navigate between sessions. I think anchoring sessions to directories is kind of genius, even if it is entirely possible to cd away from the dir you've named your session after. 

One thing i wonder is maybe if zoxide is used the right way then the order of its output (and initial fzf state) could be this very MRU listing and that could be used directly, that would be pretty elegant. But I think not since the preferred semantic is most-recently-open tmux session, that would not be something zoxide has any awareness of...

**Comment by joshmedeski** [comment]
There's a lot of questions and thoughts here @unphased! I'll do my best to cover some of what you've brought up.

1. This is all subjective, what works for me might not work for you. With that said, many people who lean into sesh love it and never look back.
2. My tmux status line is aligned to the top, combined with [gitmux](https://github.com/arl/gitmux) and my [nerd font window name plugin](https://github.com/joshmedeski/tmux-nerd-font-window-name).
3. I lean heavy into the minimalism approach. Year-by-year I keep removing UI elements. I like how clean it looks and is distraction-free. Whatever might be important to reach for is usually one keyboard binding away.
4. Changing your mindset to `session = directory` combined with zoxide's algorithm makes any project you've ever opened ~1s away at all times. Give it a try and see if you can get used to it.
5. I use the `<prefix>+L` binding (I bind it to `⌘+l`) switch between two sessions, it's very helpful if I need a session for reference or am working on a full-stack feature (with a "sever" and "web" session opened)
6. I close sessions quick and often, and I don't try and keep up with how many sessions I have open at once, with a powerful computer it doesn't really matter. I'll run `tmux kill-server` (aliased to `tk`) if I feel I have too many sessions opened and just start over.
7. My windows are made up of neovim, dev commands, and lazygit. Each can be opened with key binding. I plan on making a video on how that works so stay-tuned in for more!

Okay wow, those are some of my thoughts. I'll be making a "why use sesh?" kind of video soon and go more in depth with the mental model and how to best leverage the tool to be more productive. Thanks for showing interest and I hope these thoughts help!

**Reply by unphased** [reply]
Very helpful. Appreciate your reply. I will gradually make my transition toward minimalism and i can already tell how nice it will be to be able to use disk directories as a reference point for having "workspaces".  

- sounds like you use windows too. how would you compare working on e.g. an app with a frontend side and backend side, would you prefer to use one session and toggle windows for this or use two sessions and use your session switch binding to flip back and forth? I assume if you ever want to do a vertical split or something it'd be easier to set up if they were already together as windows in one session?
- re: kill-server i do a similar thing... when nvim has too many buffers open already i tend to kill it and reopen, and similarly every few weeks the tmux session has too many windows and panes in it and i kill the tmux server to clean it. it's not ideal but it works dammit.
- you have bindings and roles for your windows. so i guess this means you name your windows and the bind to open those will switch to the window if it exists or create it and launch the tool if it doesn't exist, which is basically an identical paradigm to your use of sessions. I like the consistency here and I think this kind of probably answers the first part of my first bullet question. But i def still want to know how you approach tmux splits since this has always been my tmux bread and butter (i never considered using screen again when i saw tmux capable of vsplits), might be safe to assume you never ever use more than 2 or 3 panes at a time though. 
- one of the primary reasons i'm interested in zellij is that it has stacked windows/panes, which is something i find lacking in tmux as i'll often end up with multiple terminals each of which represent some little mission i went off to do, the problem of course is that leaving these things around lead to clutter. I am starting to see effective ways to make full use of all of these features (Sessions, windows, panes) and it should be really nice compared to what i've been doing for 10 years which is abuse panes. Switching to stacked/collapsed panes with e.g. zellij still leaves me with a clutter problem, just one that consumes one row for each old terminal, better than 3 rows but still far worse than 0. I should open windows to do these kinds of things in. I can browse them at any time to deal with them and they take no space. I should attempt to use split panes only when i actually want to really look back and forth at stuff, otherwise full screen should be default. 

you already gave me a decent idea. navigating through nvim buffers can be cumbersome at times because a large number of buffers invariably get loaded and I keep wanting to go back to the last file but dont want to have to recall the name of it, and it may be a lot of jumping back in the jumplist before it lands me to that file. So I'm going to make a bind to do this. I made a quick way to delete buffers which helps these situations but it's still too much work to manually delete all the clutter of opened buffers from the list... 

i already moved my tmux statusline to the top, and started to clean up the status line. But to start out i am gonna remove the fields i put in there that i never reference, and gonna try to keep the ones I do read sometimes. I am writing a hook so that if the terminal is too narrow it'll switch to a two-row statusline, that is either going to be awesome or horrible. 

thanks again and sorry for all the walls of text.

**Comment by joshmedeski** [comment]
I only switch between windows to glance at dev server logs or switch to lazygit. Otherwise I'm switching between sessions to look at files in a different project. I vary rarely type `cd` or switch directories in my workflow.

**Comment by enthusiva** [comment]
@joshmedeski is there a way to get the current session information as an argument in the startup script? For example, I want to change the title of my terminal window to display the session name so that I can quickly search for specific terminal window in window switcher applications like rofi/dmenu when I have multiple terminal windows open.

**Comment by joshmedeski** [comment]
@enthusiva sure! You can use the `display-message` tmux command to print the session name and use it however you'd like.
```sh
#!/usr/bin/env bash
CURRENT_SESSION=$(tmux display-message -p "#S")
tmux new-window -n "$CURRENT_SESSION:nvim" nvim
```

Hope it helps!

**Reply by enthusiva** [reply]
Thank you. I will give this a try. I tried to set the window title using tmux client-session-changed hook but no luck yet. I don't know what I am missing.

**Reply by unphased** [reply]
You have to make sure your terminal emulator is configured to allow title changes.

**Reply by enthusiva** [reply]
@unphased Thank you so much. It worked. 
@joshmedeski Thank you so much for your continued support for queries posted on this repo.
